File: question/neetcode150/stack/car-fleet.py
from typing import List

class Solution:
    def carFleet(self, t: int, ps: List[int], ss: List[int]) -> int:
        stack = []

        a = list(zip(ps,ss))
        a.sort(reverse=True)

        for p,s in a:
            # if stack:
            #     print((t-p)/s,(t-stack[-1][0])/stack[-1][1])
            if not stack or (t-p)/s > (t-stack[-1][0])/stack[-1][1]:
                stack.append((p,s))

        return len(stack)

File: question/neetcode150/stack/car-fleet.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int carFleet(int t, vector<int>& p, vector<int>& s) {
        int n = p.size();
        vector<pair<double, double>> ps;
        
        for(int i = 0;i < n;i++){
            ps.push_back({p[i],s[i]});
        }

        sort(ps.rbegin(), ps.rend());

        // for(auto [pos,speed]: ps){
        //     cout << pos << " " << speed << endl;
        // }

        int ans = 1;
        double ns = (t-ps[0].first)/ ps[0].second;
        for (auto [pos, speed]: ps){
            double time = (t-pos)/ speed;

            // cout << t << " " << pos << " " << speed << endl;
            // cout << time << endl;
            if (ns < time){
                ns = time;
                ans++;
            }
        }

        return ans;
    }
};